File: app/src/main/java/hibernate/v2/testyourandroid/ui/fragment/ToolSoundMeterFragment.java
package hibernate.v2.testyourandroid.ui.fragment;

import android.Manifest;
import android.graphics.Color;
import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.GraphViewSeries;
import com.jjoe64.graphview.GraphViewStyle;
import com.jjoe64.graphview.LineGraphView;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import hibernate.v2.testyourandroid.C;
import hibernate.v2.testyourandroid.R;

/**
 * Created by himphen on 21/5/16.
 */
public class ToolSoundMeterFragment extends BaseFragment {

	protected final String[] PERMISSION_NAME = {Manifest.permission.RECORD_AUDIO};

	@BindView(R.id.meterCurrentTv)
	TextView meterCurrentTv;
	@BindView(R.id.meterAvgTv)
	TextView meterAvgTv;
	@BindView(R.id.meterMaxTv)
	TextView meterMaxTv;
	@BindView(R.id.meterMinTv)
	TextView meterMinTv;
	@BindView(R.id.graph2)
	LinearLayout layout;


	static final int SAMPLE_RATE_IN_HZ = 44100;

	int BUFFER_SIZE;
	short[] buffer;

	private GraphViewSeries series = null;
	private double lastXValue = 0;

	private Integer mindB = null;
	private Integer maxdB = null;
	private ArrayList<Integer> avgdB = new ArrayList<>();

	private boolean mIsRecording = false;
	private AudioRecord mAudioRecord;

	public static ToolSoundMeterFragment newInstance() {
		ToolSoundMeterFragment fragment = new ToolSoundMeterFragment();
		Bundle args = new Bundle();
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		View rootView = inflater.inflate(R.layout.fragment_tool_sound_meter, container, false);
		ButterKnife.bind(this, rootView);
		return rootView;
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		if (!isPermissionsGranted(PERMISSION_NAME)) {
			requestPermissions(PERMISSION_NAME, PERMISSION_REQUEST_CODE);
		}
	}

	@Override
	public void onPause() {
		stopRecording();
		super.onPause();
	}

	@Override
	public void onResume() {
		super.onResume();

		if (isPermissionsGranted(PERMISSION_NAME)) {
			startRecording();
		}
	}

	private void startRecording() {
		BUFFER_SIZE = AudioRecord.getMinBufferSize(SAMPLE_RATE_IN_HZ,
				AudioFormat.CHANNEL_IN_DEFAULT, AudioFormat.ENCODING_PCM_16BIT);

		if (BUFFER_SIZE < 0) {
			C.errorNoFeatureDialog(mContext);
			return;
		}

		buffer = new short[BUFFER_SIZE];
		try {
			LineGraphView graphView = new LineGraphView(mContext, "");
			graphView.setShowHorizontalLabels(false);
			series = new GraphViewSeries("", new GraphViewSeries.GraphViewSeriesStyle(
					ContextCompat.getColor(mContext, R.color.green500), 3), new GraphView.GraphViewData[]{});
			graphView.addSeries(series);
			graphView.setManualYAxisBounds(120, 0);
			graphView.setGraphViewStyle(new GraphViewStyle(Color.GRAY, Color.GRAY,
					Color.TRANSPARENT));
			graphView.setViewPort(0, 36);
			graphView.setScalable(true);
			graphView.setScrollable(true);
			graphView.setDisableTouch(true);
			layout.addView(graphView);

			mAudioRecord = new AudioRecord(
					MediaRecorder.AudioSource.MIC,
					SAMPLE_RATE_IN_HZ,
					AudioFormat.CHANNEL_IN_DEFAULT,
					AudioFormat.ENCODING_PCM_16BIT,
					AudioRecord.getMinBufferSize(
							SAMPLE_RATE_IN_HZ,
							AudioFormat.CHANNEL_IN_DEFAULT,
							AudioFormat.ENCODING_PCM_16BIT
					)
			);
			mAudioRecord.startRecording();

			mIsRecording = true;

			meterCurrentTv.post(new Runnable() {
				public void run() {
					if (mIsRecording) {
						int r = mAudioRecord.read(buffer, 0, BUFFER_SIZE);
						long v = 0;

						for (short aBuffer : buffer) {
							v += aBuffer * aBuffer;
						}

						double mean = v / (double) r;
						Integer db = Math.round(10 * (float) Math.log10(mean));

						if (db < 0) {
							db = 0;
						}

						if (maxdB == null || db > maxdB) {
							maxdB = db;
							meterMaxTv.setText(String.valueOf(maxdB));
						}

						if (mindB == null || db < mindB) {
							mindB = db;
							meterMinTv.setText(String.valueOf(mindB));
						}

						avgdB.add(db);

						meterAvgTv.setText(String.valueOf(Math.round(C.calculateAverage(avgdB))));
						meterCurrentTv.setText(String.valueOf(db));

						lastXValue += 0.5;
						series.appendData(new GraphView.GraphViewData(lastXValue, db),
								true, 100);

						if (db > 100) {
							series.getStyle().color = ContextCompat.getColor(mContext, R.color.pink500);
						} else if (db > 80) {
							series.getStyle().color = ContextCompat.getColor(mContext, R.color.gold);
						} else {
							series.getStyle().color = ContextCompat.getColor(mContext, R.color.green500);
						}

						meterCurrentTv.postDelayed(this, 500);
					}
				}
			});
		} catch (Exception e) {
			C.logException(e);
			C.errorNoFeatureDialog(mContext);
		}
	}

	@SuppressWarnings("ResultOfMethodCallIgnored")
	private void stopRecording() {
		if (mAudioRecord != null) {
			if (mIsRecording) {
				mIsRecording = false;
				try {
					mAudioRecord.stop();
				} catch (Exception ignored) {
				}
			}

			mAudioRecord.release();
			mAudioRecord = null;
		}
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults);
		if (requestCode == PERMISSION_REQUEST_CODE) {
			if (!hasAllPermissionsGranted(grantResults)) {
				C.openErrorPermissionDialog(mContext);
			}
		}
	}
}

File: app/src/main/java/hibernate/v2/testyourandroid/ui/tool/ToolSoundMeterFragment.kt
package hibernate.v2.testyourandroid.ui.tool

import android.Manifest
import android.graphics.Color
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import com.blankj.utilcode.util.ConvertUtils
import com.jjoe64.graphview.GridLabelRenderer
import com.jjoe64.graphview.series.DataPoint
import com.jjoe64.graphview.series.LineGraphSeries
import hibernate.v2.testyourandroid.R
import hibernate.v2.testyourandroid.databinding.FragmentToolSoundMeterBinding
import hibernate.v2.testyourandroid.ui.base.BaseFragment
import hibernate.v2.testyourandroid.util.Utils.errorNoFeatureDialog
import hibernate.v2.testyourandroid.util.Utils.logException
import hibernate.v2.testyourandroid.util.ext.format
import java.util.ArrayList
import kotlin.math.log10

/**
 * Created by himphen on 21/5/16.
 */
class ToolSoundMeterFragment : BaseFragment<FragmentToolSoundMeterBinding>() {

    override fun getViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): FragmentToolSoundMeterBinding =
        FragmentToolSoundMeterBinding.inflate(inflater, container, false)

    private var series = LineGraphSeries(arrayOf<DataPoint>())
    private var lastXValue = 0.0
    private var mindB: Int? = null
    private var maxdB: Int? = null
    private val avgdB = ArrayList<Int>()
    private var mIsRecording = false
    private var mAudioRecord: AudioRecord? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (!isPermissionsGranted(PERMISSION_NAME)) {
            requestMultiplePermissions.launch(PERMISSION_NAME)
        }
    }

    override fun onPause() {
        stopRecording()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        if (isPermissionsGranted(PERMISSION_NAME)) {
            startRecording()
        }
    }

    private fun startRecording() {
        context?.let { context ->
            BUFFER_SIZE = AudioRecord.getMinBufferSize(
                SAMPLE_RATE_IN_HZ,
                AudioFormat.CHANNEL_IN_DEFAULT, AudioFormat.ENCODING_PCM_16BIT
            )
            if (BUFFER_SIZE < 0) {
                errorNoFeatureDialog(context)
                return
            }
            val buffer = ShortArray(BUFFER_SIZE)
            try {
                series.thickness = ConvertUtils.dp2px(3f)
                series.color = ContextCompat.getColor(context, R.color.lineColor4)
                series.isDrawBackground = true
                series.backgroundColor = ContextCompat.getColor(context, R.color.lineColor4A)
                viewBinding?.graphView?.addSeries(series)
                viewBinding?.graphView?.viewport?.isYAxisBoundsManual = true
                viewBinding?.graphView?.viewport?.setMinY(0.0)
                viewBinding?.graphView?.viewport?.setMaxY(120.0)
                viewBinding?.graphView?.viewport?.isXAxisBoundsManual = true
                viewBinding?.graphView?.viewport?.setMinX(0.0)
                viewBinding?.graphView?.viewport?.setMaxX(36.0)
                viewBinding?.graphView?.viewport?.isScrollable = false
                viewBinding?.graphView?.viewport?.isScalable = false
                viewBinding?.graphView?.gridLabelRenderer?.gridColor = Color.GRAY
                viewBinding?.graphView?.gridLabelRenderer?.isHighlightZeroLines = false
                viewBinding?.graphView?.gridLabelRenderer?.isHorizontalLabelsVisible = false
                viewBinding?.graphView?.gridLabelRenderer?.padding = ConvertUtils.dp2px(10f)
                viewBinding?.graphView?.gridLabelRenderer?.gridStyle =
                    GridLabelRenderer.GridStyle.HORIZONTAL
                mAudioRecord = AudioRecord(
                    MediaRecorder.AudioSource.MIC,
                    SAMPLE_RATE_IN_HZ,
                    AudioFormat.CHANNEL_IN_DEFAULT,
                    AudioFormat.ENCODING_PCM_16BIT,
                    AudioRecord.getMinBufferSize(
                        SAMPLE_RATE_IN_HZ,
                        AudioFormat.CHANNEL_IN_DEFAULT,
                        AudioFormat.ENCODING_PCM_16BIT
                    )
                )
                mAudioRecord?.let {
                    it.startRecording()
                    mIsRecording = true
                    viewBinding?.meterCurrentTv?.post(object : Runnable {
                        override fun run() {
                            if (mIsRecording) {
                                val r = it.read(buffer, 0, BUFFER_SIZE)
                                var v: Long = 0
                                for (aBuffer in buffer) {
                                    v += aBuffer * aBuffer.toLong()
                                }
                                val mean = v / r.toDouble()
                                var db = (10 * log10(mean).toFloat()).toInt()
                                if (db < 0) {
                                    db = 0
                                }
                                if (maxdB == null || db > maxdB!!) {
                                    maxdB = db
                                    viewBinding?.meterMaxTv?.text = maxdB.toString()
                                }
                                if (mindB == null || db < mindB!!) {
                                    mindB = db
                                    viewBinding?.meterMinTv?.text = mindB.toString()
                                }
                                avgdB.add(db)
                                viewBinding?.meterAvgTv?.text = avgdB.average().format(0)
                                viewBinding?.meterCurrentTv?.text = db.toString()
                                lastXValue += 0.5
                                series.appendData(DataPoint(lastXValue, db.toDouble()), true, 100)
                                viewBinding?.graphView?.viewport?.scrollToEnd()
                                series.color = when {
                                    db > 100 -> ContextCompat.getColor(context, R.color.lineColor1)
                                    db > 80 -> ContextCompat.getColor(context, R.color.lineColor2)
                                    else -> ContextCompat.getColor(context, R.color.lineColor4)
                                }
                                series.isDrawBackground = true
                                series.backgroundColor = when {
                                    db > 100 -> ContextCompat.getColor(context, R.color.lineColor1A)
                                    db > 80 -> ContextCompat.getColor(context, R.color.lineColor2A)
                                    else -> ContextCompat.getColor(context, R.color.lineColor4A)
                                }
                                viewBinding?.meterCurrentTv?.postDelayed(this, 500)
                            }
                        }
                    })
                }
            } catch (e: Exception) {
                logException(e)
                errorNoFeatureDialog(context)
            }
        }
    }

    private fun stopRecording() {
        mAudioRecord?.let {
            if (mIsRecording) {
                mIsRecording = false
                try {
                    it.stop()
                } catch (ignored: Exception) {
                }
            }
            it.release()
            mAudioRecord = null
        }
    }

    companion object {
        const val SAMPLE_RATE_IN_HZ = 44100
        fun newInstance(): ToolSoundMeterFragment {
            val fragment = ToolSoundMeterFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }

        val PERMISSION_NAME = arrayOf(Manifest.permission.RECORD_AUDIO)
        var BUFFER_SIZE = 0
    }
}